Wikipedia/View Controllers/ShareCard/WMFShareOptionsController.m

```diff
@@ -0,0 +1,346 @@
+    self.originView = nil;
+    [self fetchImageThenShowShareCard];
+}
+
+
+- (void)presentShareOptionsWithSnippet:(NSString*)snippet inViewController:(UIViewController*)viewController fromView:(nullable UIView*)view{
+    self.snippet = [snippet copy];
+    self.containerViewController = viewController;
+    self.originButtonItem = nil;
+    self.originView = view;
+    [self fetchImageThenShowShareCard];
+}
+
+#pragma mark - Asynchornous Fetch and Present
+
+- (void)fetchImageThenShowShareCard {
+    [[WMFImageController sharedInstance] fetchImageWithURL:[NSURL wmf_optionalURLWithString:self.article.imageURL]].then(^(WMFImageDownload* download){
+        [self showShareOptionsWithImage:download.image];
+    }).catch(^(NSError* error){
+        [self showShareOptionsWithImage:nil];
+    });
+}
+
+#pragma mark - Share Options Setup
+
+- (void)showShareOptionsWithImage:(nullable UIImage*)image {
+    
+    [self setupBackgroundView];
+
+    _shareImage = [self cardImageWithArticleImage:image];
+    
+    [self setupShareOptionsWithImage:_shareImage];
+    
+    [self presentShareOptionsView:self.shareOptions];
+}
+
+- (void)setupBackgroundView{
+    
+    UIView* containingView = self.containerViewController.view;
+
+    UIView* grayOverlay = [[UIView alloc] initWithFrame:containingView.frame];
+    grayOverlay.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.42];
+    grayOverlay.alpha           = 0.0;
+    [containingView addSubview:grayOverlay];
+    self.grayOverlay = grayOverlay;
+    UITapGestureRecognizer* tapRecognizer = [[UITapGestureRecognizer alloc]
+                                             initWithTarget:self action:@selector(respondToDimAreaTapGesture:)];
+    [grayOverlay addGestureRecognizer:tapRecognizer];
+    grayOverlay.translatesAutoresizingMaskIntoConstraints = NO;
+    [grayOverlay mas_makeConstraints:^(MASConstraintMaker *make) {
+       
+        make.edges.equalTo(containingView);
+    }];
+}
+
+- (void)setupShareOptionsWithImage:(UIImage*)image{
+    
+    WMFShareOptionsView* shareOptionsView =
+    [[[NSBundle mainBundle] loadNibNamed:@"ShareOptions" owner:self options:nil] objectAtIndex:0];
+    shareOptionsView.cardImageViewContainer.userInteractionEnabled = YES;
+    shareOptionsView.shareAsCardLabel.userInteractionEnabled       = YES;
+    shareOptionsView.shareAsTextLabel.userInteractionEnabled       = YES;
+    shareOptionsView.shareAsCardLabel.text                         = MWLocalizedString(@"share-as-image", nil);
+    shareOptionsView.shareAsTextLabel.text                         = MWLocalizedString(@"share-as-text", nil);
+    shareOptionsView.cardImageView.image                           = _shareImage;
+    
+    [self.containerViewController.view addSubview:shareOptionsView];
+    self.shareOptions = shareOptionsView;
+}
+
+#pragma mark - Create Card Image
+
+- (UIImage*)cardImageWithArticleImage:(UIImage*)image{
+    WMFShareCardViewController* cardViewController =
```

Comment:
is the end goal to eventually remove this, similar to how the view is being loaded directly from a nib above?